codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -163,67 +178,120 @@ private static int compressionLevel(int blockSize) {
     }
 
     @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        return allocateBuffer(ctx, msg, preferDirect, true);
+    }
+
+    private ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect,
+                                   boolean allowEmptyReturn) {
+        int targetBufSize = 0;
+        int remaining = msg.readableBytes() + buffer.readableBytes();
+
+        // overflow check
+        if (remaining < 0) {
+            throw new EncoderException("too much data to allocate a buffer for compression");
+        }
+
+        while (remaining > 0) {
+            int curSize = Math.min(blockSize, remaining);
+            remaining -= curSize;
+            // calculate the total compressed size of the current block (including header) and add to the total
+            targetBufSize += compressor.maxCompressedLength(curSize) + HEADER_LENGTH;
+        }
+
+        // as compressor.maxCompressedLength(Integer.MAX_VALUE) would never wrap around again to be >= 0,
```

Comment:
Can we elaborate on this comment because it depends upon a few different factors:

- MIN_BLOCK_SIZE is 64
- HEADER_LENGTH is 21
- compressor.maxCompressedLength upper bound is `length + length / 255 + 16`

So all of these constraints means that even if `remaining == Integer.MAX_INT` we cannot overflow back around to `>= 0`.